stl/inc/memory

```diff
@@ -2416,26 +2437,298 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+        _Result._Ptr          = this->_Ptr;
+        _Result._Rep          = _Ref;
+        this->_Ptr            = _Value._Ptr;
+        this->_Unlock();
+        return _Result;
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Order);
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, const shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        _Ref_count_base* _Ref = this->_LockAndLoad();
+        if (this->_Ptr == _Expected._Ptr && _Ref == _Expected._Rep) {
+            this->_Ptr = _Desired._Ptr;
+            _Desired._Incref();
+            this->_StoreAndUnlock(_Desired._Rep);
+            if (_Ref) {
+                _Ref->_Decref();
+            }
+            return true;
+        } else {
+            _Ref_count_base* _ExpectedRep = _Expected._Rep;
+            _Expected._Ptr                = this->_Ptr;
+            _Expected._Rep                = _Ref;
+            _Expected._Incref();
+            this->_Unlock();
+            if (_ExpectedRep) {
+                _ExpectedRep->_Decref();
+            }
+            return false;
+        }
+    }
+
+    constexpr atomic() noexcept = default;
+
+    atomic(const shared_ptr<_Ty> _Value) noexcept : _Base(_Value._Ptr, _Value._Rep) {
+        _Value._Incref();
+    }
+
+    atomic(const atomic&) = delete;
+    void operator=(const atomic&) = delete;
+
+    void operator=(shared_ptr<_Ty> _Value) noexcept {
+        store(_STD move(_Value));
+    }
+
+    ~atomic() {
+        _Ref_count_base* _Ref = reinterpret_cast<_Ref_count_base*>(this->_Refptr._Storage._Value);
+        if (_Ref) {
+            _Ref->_Decref();
+        }
+    }
+};
+
+template <class _Ty>
+struct atomic<weak_ptr<_Ty>> : _Atomic_ptr_base<_Ty> {
+private:
+    using _Base = _Atomic_ptr_base<_Ty>;
+
+public:
+    using value_type = weak_ptr<_Ty>;
+
+    static constexpr bool is_always_lock_free = false;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return false;
+    }
+
+    void store(const weak_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_store_memory_order(_Order);

```

Comment:
Pretty sure all my perf comments about the shared_ptr version apply here too.

Reply:
Done